consensus/src/test/kotlin/maru/consensus/ProposerSelectorImplTest.kt

```diff
@@ -0,0 +1,115 @@
+import maru.core.ext.DataGenerators
+import maru.database.BeaconChain
+import org.assertj.core.api.Assertions.assertThat
+import org.hyperledger.besu.consensus.common.bft.ConsensusRoundIdentifier
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.mockito.ArgumentMatchers.anyLong
+import org.mockito.Mockito.mock
+import org.mockito.kotlin.whenever
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+
+class ProposerSelectorImplTest {
+  private val totalValidators = 5
+  private val validators = List(totalValidators) { DataGenerators.randomValidator() }.toSet()
+  private val genesisBlockNumber = 10uL
+  private val genesisValidator = validators.first()
+  private val validatorsAddresses = TreeSet(validators.map { it.toAddress() })
+
+  private val validatorProvider =
+    object : ValidatorProvider {
+      override fun getValidatorsForBlock(blockNumber: ULong): SafeFuture<Set<Validator>> =
+        SafeFuture.completedFuture(validators)
+    }
+
+  @Test
+  fun `select proposer for genesis block`() {
+    val beaconChain = mock<BeaconChain>()
+    val config =
+      ProposerSelectorImpl.Config(
+        changeEachBlock = true,
+        genesisBlockNumber = genesisBlockNumber,
+        genesisBlockValidator = genesisValidator,
+      )
+    val proposerSelector = ProposerSelectorImpl(beaconChain, validatorProvider, config)
+    val consensusRoundIdentifier = ConsensusRoundIdentifier(genesisBlockNumber.toLong(), 0)
+
+    val result = proposerSelector.getProposerForBlock(consensusRoundIdentifier).get()
+    assertThat(result).isEqualTo(genesisValidator)
+  }
+
+  @Test
+  fun `select proposer for next block`() {
+    val config =
+      ProposerSelectorImpl.Config(
+        changeEachBlock = true,
+        genesisBlockNumber = genesisBlockNumber,
+        genesisBlockValidator = genesisValidator,
+      )
+    var prevProposer = genesisValidator
+    val returnedValidators = mutableSetOf<Validator>()
+    for (blockNumber in genesisBlockNumber + 1uL..genesisBlockNumber + totalValidators.toULong()) {
+      val beaconChain = mock<BeaconChain>()
+      val randomBeaconBlock = DataGenerators.randomBeaconBlock(blockNumber)
+      val prevBlock =
+        randomBeaconBlock.copy(
+          beaconBlockHeader = randomBeaconBlock.beaconBlockHeader.copy(proposer = prevProposer),
+        )
+      val prevSealedBlock = SealedBeaconBlock(prevBlock, emptyList())
+      whenever(beaconChain.getSealedBeaconBlock(anyLong().toULong())).thenReturn(prevSealedBlock)
+
+      val proposerSelector = ProposerSelectorImpl(beaconChain, validatorProvider, config)
+      val consensusRoundIdentifier = ConsensusRoundIdentifier(blockNumber.toLong(), 0)
+
+      val result = proposerSelector.getProposerForBlock(consensusRoundIdentifier).get()
+      assertThat(result in validators).isTrue()
+      prevProposer = result
+      returnedValidators.add(result)
+    }
+    assertThat(returnedValidators).isEqualTo(validators)
```

Comment:
My understanding is that thanks to `changeEachBlock` out of `totalValidators` calls with consecutive block number we're getting 5 different validators. But I'm not sure if we're just lucky that the order of `returnedValidators` matches `validators` or it's the required property of `ProposerSelectorImpl`.

My understanding is that order enforcement makes this test more brittle than necessary.

Reply:
You're right. I missed that we've got a Set on both sides